Add gnome wayland support

hedning opened the pull request:
###### Motivation for this change

This adds support for launching gnome wayland sessions from GDM and SDDM, using the upstream session file. See #39871 and #32806 for some more discussion.

Edit: Updated against master now that gnome 3.30 is merged.

###### sessionCommand support

This should work quite well, but one remaining issue is the lack of support for `displayManager.sessionCommands` and therefore also `gnome3.sessionPath`. Some options here:

- We could patch GDM, gnome-session, to run the wrapper for wayland-sessions too, not super happy about it, but might be good as it makes GDM act more like SDDM and LightDM.
- Or we could move sessionCommands into /etc/profile checking for eg. XDG_SESSION_TYPE, but that seems really ugly.

Edit: We landed on simply not supporting `sessionCommands` on wayland for now. 

###### patching gnome-session more? 

6171bc0 might be better expressed by patching `gnome-session` to always look for session files in its own `share/` folder.

###### Writing a test

Edit: `gnome3-gdm` now runs a wayland gnome session (GDM prefers the wayland session and we don't have a good way of telling it otherwise), so I ported it to work with wayland.

cc @jtojnar 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

Edit:
- Tested succesfully in a VM (at https://github.com/NixOS/nixpkgs/pull/44497/commits/a6c1ffec1db2b08c8c8188fef4c09cc78cc00621).

- Ran `gnome3`, `gnome3-gdm` and  `sddm` tests succesfully.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

hedning committed:
gnome3.gnome-session: prevent crash when launching wayland sessions

gnome-session inherits GDMS PATH, which is at the moment non-functional. In X11
this didn't matter as the `Xsession` wrapper would populate the environment
beforehand. Wayland sessions doesn't source `Xesssion` (duh), so we patch
`bin/gnome-session` to use absolute paths for `grep` and `bash`.

In addition `bin/gnome-session` is a simple wrapper around
`libexec/gnome-session-binary` mostly responsible for sourcing the users profile
before launching the binary. This made our wrapping of `bin/gnome-session`
ineffective on wayland as the profile would reset the environment. Simply wrap
`libexec/gnome-session-binary` instead.

hedning committed:
nixos/displayManager: Install wayland sessions from extraSessionFilePackages

hedning committed:
nixos/sddm: Enable wayland-sessions

LightDM is unable to separate between `wayland-sessions/gnome.desktop` and
`xsessions/gnome.desktop` so I ommitted adding this to LightDM.

jtojnar commented on pkgs/desktops/gnome-3/core/gnome-session/fix-paths.patch:
```diff
@@ -20,3 +36,42 @@
+         manager->priv = GSM_MANAGER_GET_PRIVATE (manager);
+ 
+-        manager->priv->settings = g_settings_new (GSM_MANAGER_SCHEMA);
++        GSettingsSchemaSource *schema_source = g_settings_schema_source_new_from_directory("@gschemasCompiled@", NULL, FALSE, NULL);
```
`GSettingsSchemaSource` and `GSettingsSchema` need to be freed using [`g_settings_schema_source_unref`](https://developer.gnome.org/gio/stable/gio-GSettingsSchema-GSettingsSchemaSource.html#g-settings-schema-source-unref) and [`g_settings_schema_unref`](https://developer.gnome.org/gio/stable/gio-GSettingsSchema-GSettingsSchemaSource.html#g-settings-schema-unref) respectively.

hedning replied:
Tried freeing right after `manager->priv->settings` is assigned, and that seemed to work. Though I can see that you've waited to free until the settings object is freed in the gpaste patch, so I'm not entirely sure if it actually works.

If `settings` hold a reference to the `schema` and `schema_source` I'm guessing it will correctly free it when destroyed?

jtojnar replied:
They are reference counted so if `settings` holds the reference it will unref it, otherwise there should be no issue.

hedning replied:
The VM worked at least, so going with that.

jtojnar replied:
Yeah, got a confirmation on IRC.

* https://gitlab.gnome.org/GNOME/glib/blob/516101c7023af25083c385b91a1ca7dc635c15a6/gio/gsettingsschema.c#L177
* https://gitlab.gnome.org/GNOME/glib/blob/516101c7023af25083c385b91a1ca7dc635c15a6/gio/gsettings.c#L548
* https://gitlab.gnome.org/GNOME/glib/blob/516101c7023af25083c385b91a1ca7dc635c15a6/gio/gsettings.c#L716

hedning committed:
nixos/gnome3: Implement `sessionPath` through `environment.extraInit`

This will simply make the `sessionPath` more likely to work.

hedning committed:
nixos/displayManager: Note that `sessionCommands` aren't run on Wayland

hedning commented:
Replying to https://github.com/NixOS/nixpkgs/issues/39871#issuecomment-410655127 here instead.

> I would not call the xsessionWrapper particularly elegant solution either. I would like to use the Xsession provided by GDM in the end.

Agreed, and in Gnome's case it seems the session is set up correctly with eg. pulseaudio, journal logging and dbus without the wrapper, mostly leaving `/etc/profile` which is sourced correctly. The only missing things seems to be `sessionsCommands` and `graphical-session.target` I think.

Any idea how to get the `graphical-session.target` running?

> For what is worth, sessionPath might actually make sense outside of GNOME (e.g. for elementary), so moving it to /etc/profile should work.

True, there's not really any reason to switch on XDG_CURRENT_SESSION, users probably want functional software if they happen to currently run something else than Gnome. I'm guessing it even might make sense in a console session, as there could be cli tools requiring something exposed in these variables (probably not likely, but it shouldn't hurt).

> As for sessionCommands, I would just add a comment it is not supported for sessions created from extraSessionFilePackages.

I'm not opposed to that. It should be noted that `sessionCommands` are currently sourced for X sessions from extraSessionFilePackages, due to the `xsessionWrapper`, so it's only wayland sessions that won't run it, which is even less breakage :)

##### Default session

One more problem, the wayland session have the same name as the xorg session (there's an extra session file that's explicitly for xorg) and seem to take priority. So people might end up «upgrading» to wayland without any warning which could be jarring.

jtojnar commented:
> Any idea how to get the `graphical-session.target` running?

Not sure. These might be relevant:

* https://gitlab.gnome.org/GNOME/gdm/issues/396
* https://gitlab.gnome.org/GNOME/rygel/issues/106

> True, there's not really any reason to switch on XDG_CURRENT_SESSION, users probably want functional software if they happen to currently run something else than Gnome.

For GNOME Flashback and Elementary, I have changed it to https://github.com/jtojnar/nixpkgs/commit/b5b2d8d92ec95c30068457108421ac5f31c841c1 but removal of the condition might make sense too.

> I'm not opposed to that. It should be noted that `sessionCommands` are currently sourced for X sessions from extraSessionFilePackages, due to the `xsessionWrapper`.

Right. I would actually expect most of the things in the `xsessionWrapper` to be in the `xsession` instead. After all, upstream sessions should be launching most of these themselves, there should only be the Xorg things, see `${gnome3.gdm}/etc/gdm/Xsession`.

> So people might end up «upgrading» to wayland without any warning which could be jarring.

I would not worry about that. If the Wayland session fails to start for some reason, `gnome-xorg` will be selected automatically.

Actually, I am not sure GDM even respects default DE nixos option. I think it uses the user’s last selected one.

hedning commented:
Hmm, so looks like we're abusing `graphical-session.target`. In that case I'm at least fine with breaking that in the wayland session.

> Right. I would actually expect most of the things in the xsessionWrapper to be in the xsession instead. After all, upstream sessions should be launching most of these themselves, there should only be the Xorg things, see ${gnome3.gdm}/etc/gdm/Xsession.

Hmm, true, I didn't think that through fully when I wrote it. The wrapper should model gdm's `Xsession` more closely, though `${sessionCommands}` will need to go there too. Perhaps out of scope of this PR, though could of course put it here anyway if it's wanted?

hedning commented:
Another issue I bumped into (dogfooding the PR at the moment):

`gnome-session` launches a login shell using the user's preferred shell (as opposed to `xsessionWrapper` sourcing `/etc/profile`). In my case ZSH.

This results in any new interactive bash shells to source `/etc/profile`. In particular running `nix run nixpkgs.some-package` will result in an  environment with a PATH that doesn't include the package. `nix-shell` is unaffected, so this might be a bug in `nix`, but it's never the less very sub-optimal. We're basically hitting a strange variant of #33219

The problem boils down to [`$__ETC_PROFILE_DONE`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/bash/bash.nix#L188) never being set.

In lieu of `/etc/environment` we would probably need some sort of shell agnostic variant of `__ETC_PROFILE_DONE` that guards against sourcing the environment again for interactive shells.

bkchr commented:
Any update on this pull request?

jtojnar commented:
This is blocked by https://github.com/NixOS/nixpkgs/pull/45950 which is blocked by https://github.com/NixOS/nixpkgs/pull/46020

nathyong commented:
What's the status on this now that gnome3 has been updated?

hedning commented:
@nathyong gnome-3.30 is in [staging-next](https://github.com/NixOS/nixpkgs/pull/51444), I'll fix up this PR when it hits master, then we should have wayland support pretty soon :)

hedning committed:
nixos/tests/gnome3: select X11 gnome shell explicitely

This isn't strictly necessary yet as LightDM doesn't read the wayland sessions,
but there's no harm in being explicit.

hedning committed:
nixos/tests/gnome3-gdm: port to wayland

The test now runs wayland, which means we can no longer use X11 style testing.
Instead we get gnome shell to execute javascript through its dbus interface.

hedning commented:
Ported the `gnome3-gdm` test to wayland (it broke since gdm prefers the wayland session. It's not super pretty, but should work, so we'll have test too :)

I think this should be ready now :tada:

hedning commented:
Found a much simpler fix for `gnome-session`: simply wrap `libexec/gnome-session-binary` instead of `bin/gnome-session`, that way sourcing the profile in `gnome-session` won't clobber the wrapping environment.

Reran the tests, will merge later today if no one has any objections.

hedning commented:
@GrahamcOfBorg build nixosTests.gnome3-gdm

hedning commented:
The test timed out on ofborg, hopefully not due to testing for active gnome shell too early, putting too much pressure on the machine. Lets see how it dose on hydra.
